staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go

```diff
@@ -77,4 +79,33 @@ type Response struct {
 	Audiences Audiences
 	// User is the UserInfo associated with the authentication context.
 	User user.Info
+	// AuthenticatorID identifies the authenticator that authenticated the request.
+	AuthenticatorID string
+}
+
+// Error is the error returned by an authenticator upon failures.
+type Error struct {
+	// Err is the error produced by an authenticator.
+	Err error
+	// AuthenticatorID identifies the authenticator that failed an authentication request.
+	AuthenticatorID string
+}
+
+// Error implements the errors.Error interface.
+func (a *Error) Error() string {
+	return fmt.Sprintf("authenticator: %q failed to process authentication request: err: %v", a.AuthenticatorID, a.Err)
+}
+
+// NewError constructs Error.
+// When the supplied error is nil, this method will also return nil.
+func NewError(authenticatorID string, err error) *Error {
+	if err == nil {
+		return nil
+	}
+	return &Error{AuthenticatorID: authenticatorID, Err: err}
```

Comment:
Why is this a pointer? Just have Error implement error.

Reply:
@mikedanese 
Error does implement error already.

The reason for using a pointer was to handle the scenario where the caller supplied nil in the input for the error. 
In other words, I was trying to avoid the scenario where Error{AuthenticatorName: "foo", Err: nil} is returned, since this does not make much sense.

If we are OK to assume that the caller never supplies Nil, then getting rid of the pointer would make sense. I am open to suggestions here.